health/reducers/transcode.go

```diff
@@ -44,24 +44,31 @@ func (t TranscodeReducer) Conditions() []health.ConditionType {
 	return transcodeConditions
 }
 
-func (t TranscodeReducer) Reduce(current health.Status, _ interface{}, evtIface data.Event) (health.Status, interface{}) {
+func (t TranscodeReducer) Reduce(current *health.Status, _ interface{}, evtIface data.Event) (*health.Status, interface{}) {
 	evt, ok := evtIface.(*data.TranscodeEvent)
 	if !ok {
 		return current, nil
 	}
 
 	ts := evt.Timestamp()
 	conditions := make([]*health.Condition, len(current.Conditions))
-	for i, cond := range current.Conditions {
-		status := conditionStatus(evt, cond.Type)
-		if status == nil {
-			conditions[i] = cond
-			continue
+	copy(conditions, current.Conditions)
+	for i, cond := range conditions {
+		if status := conditionStatus(evt, cond.Type); status != nil {
+			conditions[i] = health.NewCondition(cond.Type, ts, status, nil, cond)
 		}
-		conditions[i] = health.NewCondition(cond.Type, ts, status, nil, cond)
 	}
 
-	return health.Status{
+	// This literal is breaking multistream because it doesn't copy the
+	// multistream field. Doesn't seem scalable to keep having to update all
+	// literals for copies so
+	//
+	// TODO: Create some kind of `.Clone()` function in `Status` to create a deep
+	// copy of the object that we can mutate freely in the pipeline. Consider
+	// calling it directly from `Core` and do all the mutations in place in these
+	// reducers (although they would stop being "reducers" per se, maybe better to
+	// keep copying on each of them)
```

Comment:
Thinking furtger about it, I think deep copying in the beginning of the pipeline, in `health.Core`, and then just mutating them in place here will result in some nicer code as well as easier to maintain invariants in the code.

This "never mutate status in place" invariant is a bit tricky, and I was already transgressing it in some places like the stats reducer where I knew in the current code it was safe, but this means that the design is not very round up.

So my current mind is on:
 - create a "deep copy" helper in status struct
 - use it both when returning a status outside of the `Core` (`GetStatus()`) as well as in the beginning of the reducers pipeline (optimal locking)
 - reducers mutate the status in place, not worrying about copying a bunch of stuff around for every little update. Don't even need to return a `Status` object, only state, which will make it clearer how the update should be done.

Should be a pretty quick change as well, so I think it's a small refactor worth getting in this PR (which fixes the TODO here as well)

Reply:
Hmmm. My gut says keeping things immutable would be nice. Thinking of like... events coming through the pipeline that won't necessarily change a particular stream, but _might_ do so? We wouldn't want to create a deep-copy unnecessarily, especially because we want to optimize this for throughput... but I'm sympathetic, as I can't think of a really clean way to implement this in Go. I guess you could have a merge function that takes an old Status and a partial Status and returns a new merged one — or the old one if there weren't any changes. Not that bad, but it's a shame you'd have to do it for every reducer-value type.

Reply:
I do like the immutability as well, even if soft, so I prefer to keep it for now until we decide it really won't work anymore. So I did a "partial" solution for this here: https://github.com/livepeer/livepeer-data/pull/7/commits/1e05519177e88e67fda1f2b704477450980c9085

Basically created a couple helpers to make the copying of slices and the merging of the `Status` struct a bit nicer. The merge helper ended up exactly like you suggested, "taking an old Status and a partial Status and returning a new merged one".

Also added some comments in the struct so this "immutability" is not only implicit in the code.

Reply:
> Hmmm. My gut says keeping things immutable would be nice. 

doesn't play good with

>  optimize this for throughput

go lang's biggest performance problem is the memory allocation. So if we really want to optimize for throughput we should avoid memory allocations as a plague.